estransport/estransport.go

```diff
@@ -84,41 +85,90 @@ func New(cfg Config) *Client {
 //
 func (c *Client) Perform(req *http.Request) (*http.Response, error) {
 	var (
-		dupReqBody io.Reader
+		res *http.Response
+		err error
+
+		dupReqBodyForLog io.ReadCloser
+
+		// TODO(karmi): Make dynamic based on number of URLs
+		maxRetries = defMaxRetries
 	)
 
-	// Get URL from the Selector
+	// TODO: Handle context deadline
 	//
-	u, err := c.getURL()
-	if err != nil {
-		// TODO(karmi): Log error
-		return nil, fmt.Errorf("cannot get URL: %s", err)
-	}
+	// Must be run in a goroutine in order to not block the entire operation?
+	//
+	// if req.Context != nil {
+	// 	select {
+	// 	// Return immediately when context timeout is exceeded
+	// 	case <-req.Context().Done():
+	// 		return nil, req.Context().Err()
+	// 	default:
+	// 	}
+	// }
 
 	// Update request
 	//
-	c.setURL(u, req)
 	c.setUserAgent(req)
-	c.setAuthorization(u, req)
 
-	// Duplicate request body for logger
-	//
-	if c.logger != nil && c.logger.RequestBodyEnabled() {
-		if req.Body != nil && req.Body != http.NoBody {
-			dupReqBody, req.Body, _ = duplicateBody(req.Body)
+	for i := 1; ; i++ {
+		var nodeURL *url.URL
+
+		// fmt.Printf("\nPerform: attempt %d\n", i)
+
+		// Get URL from the Selector
+		//
+		nodeURL, err = c.getURL()
+		if err != nil {
+			// TODO(karmi): Log error
+			return nil, fmt.Errorf("cannot get URL: %s", err)
 		}
-	}
 
-	// Set up time measures and execute the request
-	//
-	start := time.Now().UTC()
-	res, err := c.transport.RoundTrip(req)
-	dur := time.Since(start)
+		// Update request
+		//
+		c.setURL(nodeURL, req)
+		c.setAuthorization(nodeURL, req)
+
+		// Duplicate request body for logger
+		//
+		if c.logger != nil && c.logger.RequestBodyEnabled() {
+			if req.Body != nil && req.Body != http.NoBody {
+				dupReqBodyForLog, req.Body, _ = duplicateBody(req.Body)
```

Comment:
You can use `req.GetBody` here:

```go
if req.Body != nil && req.Body != http.NoBody {
    body, err := req.GetBody()
    if err != nil {
        return nil, err
    }
    dupReqBodyForLog = body
}
```

Or you could just let `logRoundTrip` do that.